### PR TITLE
BUG: Accept calRGB and calGray color_spaces

### LINKED ISSUE #1956: TypeError: unhashable type: 'ArrayObject' when enumerating images
Background is the same as #1954 and #1955. Seems very similar to #1955 but I made a separate issue just in case it's different enough to need a different fix.

```
╭─────────────────────────────── Traceback (most recent call last) ────────────────────────────────╮
│ /Users/syblius/workspace/clown_sort/clown_sort/files/pdf_file.py:52 in extracted_text            │
│                                                                                                  │
│    49 │   │   │   │                                                                              │
│    50 │   │   │   │   # Extracting images is a bit fraught (lots of PIL and pypdf exceptions h   │
│    51 │   │   │   │   try:                                                                       │
│ ❱  52 │   │   │   │   │   for image_number, image in enumerate(page.images, start=1):            │
│    53 │   │   │   │   │   │   image_name = f"Page {page_number}, Image {image_number}"           │
│    54 │   │   │   │   │   │   self._log_to_stderr(f"   Processing {image_name}...")              │
│    55 │   │   │   │   │   │   page_buffer.print(Panel(image_name, expand=False))                 │
│                                                                                                  │
│ /Users/syblius/Library/Caches/pypoetry/virtualenvs/clown-sort-zLqmJuxs-py3.11/lib/python3.11/sit │
│ e-packages/pypdf/_page.py:2596 in __iter__                                                       │
│                                                                                                  │
│   2593 │                                                                                         │
│   2594 │   def __iter__(self) -> Iterator[ImageFile]:                                            │
│   2595 │   │   for i in range(len(self)):                                                        │
│ ❱ 2596 │   │   │   yield self[i]                                                                 │
│   2597 │                                                                                         │
│   2598 │   def __str__(self) -> str:                                                             │
│   2599 │   │   p = [f"Image_{i}={n}" for i, n in enumerate(self.ids_function())]                 │
│                                                                                                  │
│ /Users/syblius/Library/Caches/pypoetry/virtualenvs/clown-sort-zLqmJuxs-py3.11/lib/python3.11/sit │
│ e-packages/pypdf/_page.py:2592 in __getitem__                                                    │
│                                                                                                  │
│   2589 │   │   │   index = len_self + index                                                      │
│   2590 │   │   if index < 0 or index >= len_self:                                                │
│   2591 │   │   │   raise IndexError("sequence index out of range")                               │
│ ❱ 2592 │   │   return self.get_function(lst[index])                                              │
│   2593 │                                                                                         │
│   2594 │   def __iter__(self) -> Iterator[ImageFile]:                                            │
│   2595 │   │   for i in range(len(self)):                                                        │
│                                                                                                  │
│ /Users/syblius/Library/Caches/pypoetry/virtualenvs/clown-sort-zLqmJuxs-py3.11/lib/python3.11/sit │
│ e-packages/pypdf/_page.py:522 in _get_image                                                      │
│                                                                                                  │
│    519 │   │   │   │   │   raise KeyError("no inline image can be found")                        │
│    520 │   │   │   │   return self.inline_images[id]                                             │
│    521 │   │   │                                                                                 │
│ ❱  522 │   │   │   imgd = _xobj_to_image(cast(DictionaryObject, xobjs[id]))                      │
│    523 │   │   │   extension, byte_stream = imgd[:2]                                             │
│    524 │   │   │   f = ImageFile(                                                                │
│    525 │   │   │   │   name=f"{id[1:]}{extension}",                                              │
│                                                                                                  │
│ /Users/syblius/Library/Caches/pypoetry/virtualenvs/clown-sort-zLqmJuxs-py3.11/lib/python3.11/sit │
│ e-packages/pypdf/filters.py:844 in _xobj_to_image                                                │
│                                                                                                  │
│   841 │   filters = x_object_obj.get(SA.FILTER, [None])                                          │
│   842 │   lfilters = filters[-1] if isinstance(filters, list) else filters                       │
│   843 │   if lfilters == FT.FLATE_DECODE:                                                        │
│ ❱ 844 │   │   img, image_format, extension = _handle_flate(                                      │
│   845 │   │   │   size, data, mode, color_space, colors                                          │
│   846 │   │   )                                                                                  │
│   847 │   elif lfilters in (FT.LZW_DECODE, FT.ASCII_85_DECODE, FT.CCITT_FAX_DECODE):             │
│                                                                                                  │
│ /Users/syblius/Library/Caches/pypoetry/virtualenvs/clown-sort-zLqmJuxs-py3.11/lib/python3.11/sit │
│ e-packages/pypdf/filters.py:741 in _handle_flate                                                 │
│                                                                                                  │
│   738 │   │   │   │   │   │   "P": (0, "", ""),                                                  │
│   739 │   │   │   │   │   │   "RGB": (3, "P", "RGB"),                                            │
│   740 │   │   │   │   │   │   "CMYK": (4, "P", "CMYK"),                                          │
│ ❱ 741 │   │   │   │   │   }[_get_imagemode(base, 0, "")]                                         │
│   742 │   │   │   │   except KeyError:  # pragma: no cover                                       │
│   743 │   │   │   │   │   logger_warning(                                                        │
│   744 │   │   │   │   │   │   f"Base {base} not coded please share the pdf file with pypdf dev   │
│                                                                                                  │
│ /Users/syblius/Library/Caches/pypoetry/virtualenvs/clown-sort-zLqmJuxs-py3.11/lib/python3.11/sit │
│ e-packages/pypdf/filters.py:682 in _get_imagemode                                                │
│                                                                                                  │
│   679 │   │   "/DeviceCMYK": "CMYK",                                                             │
│   680 │   }                                                                                      │
│   681 │   mode: mode_str_type = (                                                                │
│ ❱ 682 │   │   mode_map.get(color_space)  # type: ignore                                          │
│   683 │   │   or list(mode_map.values())[color_components]                                       │
│   684 │   │   or prev_mode                                                                       │
│   685 │   )  # type: ignore                                                                      │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
TypeError: unhashable type: 'ArrayObject'
```

### PR DESCRIPTION
**Pull request by pubpub-zz**
Fixes #1956

### CODE REVIEW

**Comment by MartinThoma** [comment]
I've just noticed that there is an issue on `main` which I fix in https://github.com/py-pdf/pypdf/pull/1969

File: pypdf/filters.py [hunk 1]
```diff
@@ -654,6 +654,9 @@ def _get_imagemode(
         raise PdfReadError(
             "can not interprete colorspace", color_space
         )  # pragma: no cover
+    elif color_space[0].startswith("/Cal"):  # /CalRGB and /CalGray
+        color_space = "/Device" + color_space[0][4:]
+        # color_components =
```
Comment (MartinThoma) [inline pypdf/filters.py:659]:
```suggestion
```

**Comment by MartinThoma** [comment]
@pubpub-zz This one looks good to me :-) Can I merge?

**Comment by pubpub-zz** [comment]
All good

File: tests/test_filters.py [hunk 2]
```diff
@@ -456,4 +464,4 @@ def test_2bits_image():
     d = sqrt(
         sum([(a * a + b * b + c * c + d * d) for a, b, c, d in diff.getdata()])
     ) / (diff.size[0] * diff.size[1])
-    assert d < 0.01
+    assert d < 0.01
```
Comment (MartinThoma) [inline tests/test_filters.py:467]:
```suggestion
    assert d < 0.01

```

**Comment by MartinThoma** [comment]
Thank you :tada:

**Pull request merged**